types/src/transaction/webauthn.rs

```diff
@@ -0,0 +1,693 @@
+        .copied()
+        .collect::<Vec<u8>>()
+}
+
+/// This checks that the SHA3-256 of the `signing_message(message)`
+/// is equal to the actual `challenge` from `client_data_json` in `PartialAuthenticatorAssertionResponse`
+fn verify_expected_challenge_from_message_matches_actual<T: Serialize + CryptoHash>(
+    message: &T,
+    actual_challenge: &[u8],
+) -> std::result::Result<bool, CryptoMaterialError> {
+    // Generate signing_message
+    let signing_message_bytes = signing_message(message)?;
+    // Expected challenge is SHA3-256 digest of RawTransaction bytes
+    let expected_challenge = HashValue::sha3_256_of(signing_message_bytes.as_slice());
+
+    // Handle invariant where lengths are different
+    if actual_challenge.len() != expected_challenge.to_vec().len() {
+        return Err(CryptoMaterialError::WrongLengthError);
+    }
+
+    // Compute deep equal of actual challenge and message (expected challenge)
+    let deep_equal_challenge = actual_challenge
+        .iter()
+        .zip(expected_challenge.to_vec().iter())
+        .all(|(a, b)| a == b);
+
+    // Verify that message is equal to challenge
+    if deep_equal_challenge {
+        return Ok(deep_equal_challenge);
+    }
+    Err(CryptoMaterialError::ValidationError)
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum WebAuthnSignature {
+    P256Ecdsa { signature: p256_ecdsa::Signature },
+}
+
+/// `PartialAuthenticatorAssertionResponse` includes a subset of the fields returned from
+/// an [`AuthenticatorAssertionResponse`](passkey_types::webauthn::AuthenticatorAssertionResponse)
+///
+/// See <https://www.w3.org/TR/webauthn-3/#authenticatorassertionresponse>
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub struct PartialAuthenticatorAssertionResponse {
+    /// This attribute contains the raw signature returned from the authenticator.
+    /// NOTE: Many signatures returned from WebAuthn assertions are not raw signatures.
+    /// As an example, secp256r1 signatures are encoded as an [ASN.1 DER Ecdsa-Sig_value](https://www.w3.org/TR/webauthn-3/#sctn-signature-attestation-types)
+    /// If the signature is encoded, the client is expected to convert the encoded signature
+    /// into a raw signature before including it in the transaction
+    signature: WebAuthnSignature,
+    /// This attribute contains the authenticator data returned by the authenticator.
+    /// See [`AuthenticatorData`](passkey_types::ctap2::AuthenticatorData).
+    #[serde(with = "serde_bytes")]
+    authenticator_data: Vec<u8>,
+    /// This attribute contains the JSON byte serialization of [`CollectedClientData`](CollectedClientData) passed to the
+    /// authenticator by the client in order to generate this credential. The exact JSON serialization
+    /// MUST be preserved, as the hash of the serialized client data has been computed over it.
+    #[serde(with = "serde_bytes")]
+    client_data_json: Vec<u8>,
```

Comment:
I only see the challenge field being used, do we need the whole json here?

Reply:
unfortunately, I believe so. We need to prove a couple things during verification:
1. client_data contains the expected challenge
2. signature is over `authenticator_data || Hash(client_data)`

I'm not sure how you'd do both without including the client_data. Is there something you see that I'm missing?

Reply:
As we just discussed earlier today, if `client_data_json` is canonically-encoded, then you can reconstruct it from the TXN hash and the rest of its fields. This will make your signature size smaller and avoid parsing/deserialization bugs when matching the challenge inside the JSON.

Reply:
I... worry about manipulating client_data_json here. They use a custom JSON serialization algorithm, specified here in [5.8.1.1 of the WebAuthn spec](https://www.w3.org/TR/webauthn-3/#clientdatajson-serialization) and I'd rather not mess with it. If that serialization algorithm changes in any way in the future, and we don't include the raw client_data_json bytes returned from a WebAuthn authenticator assertion response, we'll have no way of reconstructing the original message `m` -- `authenticator_data || SHA-256(client_data_json)`.

Reply:
I think what you are saying is: it is not canonically encoded so we cannot easily reconstruct it 🙂.